cmd/ingester/app/consumer/deadlock_detector.go

```diff
@@ -82,18 +84,23 @@ func newDeadlockDetector(metricsFactory metrics.Factory, logger *zap.Logger, int
 func (s *deadlockDetector) startMonitoringForPartition(partition int32) *partitionDeadlockDetector {
 	var msgConsumed uint64
 	w := &partitionDeadlockDetector{
-		msgConsumed:    &msgConsumed,
-		partition:      partition,
-		closePartition: make(chan struct{}, 1),
-		done:           make(chan struct{}),
-		logger:         s.logger,
+		msgConsumed: &msgConsumed,
+		partition:   partition,
+		logger:      s.logger,
+		disabled:    0 == s.interval,
 
 		incrementAllPartitionMsgCount: func() {
 			s.allPartitionsDeadlockDetector.incrementMsgCount()
 		},
 	}
 
-	go s.monitorForPartition(w, partition)
+	if s.interval == 0 {
+		s.logger.Debug("Partition deadlock detector disabled")
+	} else {
+		w.closePartition = make(chan struct{}, 1)
+		w.done = make(chan struct{})
```

Comment:
nit: I don't think it's necessary to pull these out of the main construction. It saves a couple of memory allocations, but makes the code harder to maintain.